### PR TITLE
sc-3751 Add sequence CRDT

### PR DESCRIPTION
**Pull request by pdeziel**
This adds a "sequence" CRDT, which is the data structure that keeps track of insert/append/remove operations across all the nodes. This makes it possible to have a consistent state of a sequence of objects across replicas. Since there is a defined ordering for all of the operations, the sequences across replicas converge to the same state when they are merged together. This data structure could be used to represent a "cell" in a Python notebook.

### CODE REVIEW

File: crdt/sequence.py [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+        """
+        if position < 0 or position >= len(self.sequence):
+            raise IndexError("Index out of bounds")
+
+        index = -1
+        while position >= 0:
+            index += 1
+            if not self.sequence[index].tombstone:
+                position -= 1
+        return index
```
Comment (pdeziel) [inline crdt/sequence.py:53]: This is required so that the caller doesn't have to know anything about operation IDs when performing inserts and removals and they can just pass in a normal index.

File: crdt/sequence.py [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+        Appends an item to the end of the sequence.
+        """
+        
+        # Tick the clock
+        self.clock.add(1)
+
+        # Add the insert operation to the log and update the sequence
+        op = Operation(OpId(self.id, self.clock.get()), OperationType.INSERT, None, payload=object)
+        self.operations.add(op)
+        op.do(self.sequence)
```
Comment (pdeziel) [inline crdt/sequence.py:68]: A quirk (or feature?) of this implementation is that `append` works very differently from `insert` in terms of object ordering. `insert` means "insert before this object" and `append` means "insert at the end". This means that writes from different nodes can get interleaved on appends, while for `inserts` writes have a direct relationship to another object and writes from the same node get chunked together. I think this is why Peritext/automerge defines appends as referencing the last object instead of "put this at the end", because appends are very common in document editing.

File: crdt/sequence.py [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+        Merges another Sequence with this one.
+        """
+        if not isinstance(other, Sequence):
+            raise ValueError("Incompatible CRDT for merge(), expected Sequence")
+        
+        # Sync the local clock with the remote clock
+        self.clock = self.clock.merge(other.clock)
+
+        # Get the new operations to be applied
+        patch_ops = other.operations.get().difference(self.operations.get())
```
Comment (pdeziel) [inline crdt/sequence.py:118]: We don't have to apply all the operations here, just the ones we don't know about.

File: crdt/sequence.py [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+    Operations to be imposed.
+    """
+
+    def __init__(self, owner, action, target, payload=None):
+        if action not in OperationType:
+            raise ValueError("Invalid operation type")
+        self.owner = owner
+        self.action = action
+        self.target = target
+        self.payload = payload
```
Comment (pdeziel) [inline crdt/sequence.py:244]: Let me know if the terminology is confusing, the `owner` is the originator of the operation which is an `OpId`. The target is also an `OpId` which is the operation that this operation is performed on.

File: crdt/sequence.py [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+            index = 0
+            while index < len(objects):
+                op = objects[index].operation
+                if op.target == self.target and self.owner.is_earlier(op.owner):
+                    # We found another operation older than us, so stop iteration.
+                    break
+                elif op == self.target:
+                    # We found the target index
+                    break
+                index += 1
```
Comment (pdeziel) [inline crdt/sequence.py:274]: This is the most important piece of code; it handles the merging of operations into the object list so that we can maintain the current state of the sequence. Since we are inserting operations after the fact (due to getting unseen operations from the remote), this requires a scan of the object list which in the worst case we have to scan the entire operation set. There is room for optimization here; maybe we could also discard tombstoned operations periodically so the search space doesn't get unwieldy.

File: crdt/sequence.py [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+    def __init__(self, id=uuid.uuid4()):
+        self.id = id
+        self.operations = GSet()
+        self.clock = GCounter(self.id)
+        self.sequence = []
+
+    def compare_operations(self, a, b):
+        """
+        Compares this Operation to another Operation to determine ordering.
+        """
```
Comment (rebeccabilbro) [inline crdt/sequence.py]: Would you mind adding to the docstring what types `a` and `b` are here, as well as the expected return type?

File: crdt/sequence.py [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+        return a.owner.compare(b.owner)
+
+    def index_transform(self, position):
+        """
+        Returns the transformed index, which is the actual index of the object in the
+        sequence, taking into account tombstones.
+        """
```
Comment (rebeccabilbro) [inline crdt/sequence.py]: Might also be useful to document the input and output here; something like:

`position`: int
    the place where the user intends to insert, but which does not take into consideration items that have been removed by prior operations (aka tombstones)
`index`: int
     the correct insertion place, taking into account tombstones

File: crdt/sequence.py (hunk 2 above)
Comment (rebeccabilbro) [reply crdt/sequence.py:68]: I recommend adding the above comment to the docstring

File: crdt/sequence.py [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+        Prints the object to stdout.
+        """
+        print("owner: {}, payload: {}, tombstone: {}".format(self.operation.owner.print(), self.operation.payload, self.tombstone))
+
+class OpId():
+    """
+    An OpId represents a unique identifier for an Operation.
+    """
+
+    def __init__(self, name, id):
```
Comment (rebeccabilbro) [inline crdt/sequence.py]: I recommend adding a docstring to explain what types `name` and `id` are and how they should be used, particularly since they are both used to determine precedence, but differently.

File: crdt/sequence.py [hunk 9]
```diff
@@ -0,0 +1,247 @@
...
+class Object():
+    """
+    An Object represents a single item in a Sequence.
+    """
+    
+    def __init__(self, operation):
+        self.operation = operation
+        self.tombstone = False
+
+    def print(self):
```
Comment (rebeccabilbro) [inline crdt/sequence.py]: Nice! Alternatively you could overload `__repr__` (similar to how you overload `__eq__` below), which is another Python magic method:

```python
def __repr__(self):
```

File: crdt/sequence.py (hunk 4 above)
Comment (rebeccabilbro) [reply crdt/sequence.py:244]: Why don't we go ahead and add this to the docstring to make sure we remember what type each param is and how they're intended to be used. Also, would you mind making them all keyword arguments rather than positional; perhaps they can all default to `None`?

File: crdt/sequence.py [hunk 10]
```diff
@@ -0,0 +1,247 @@
...
+    """
+    Sequence is a CRDT that represents an ordered set of objects and supports insertion
+    and removal at arbitrary positions in the set.
+    """
+
+    def __init__(self, id=uuid.uuid4()):
+        self.id = id
+        self.operations = GSet()
+        self.clock = GCounter(self.id)
+        self.sequence = []
```
Comment (rebeccabilbro) [inline crdt/sequence.py:18]: What do you think about also keeping a count of stomps, just for fun? It might take a small amount of tweaking to `Operation.do` but might be a neat feature!

File: crdt/sequence.py [hunk 11]
```diff
@@ -0,0 +1,247 @@
...
+        self.target = target
+        self.payload = payload
+
+    def do(self, objects):
+        """
+        Applies this Operation to an ordered list of objects.
+        """
+        # Iterate through the sequence to find the index of the target object
+        stomp = False
+        if self.target is None:
```
Comment (rebeccabilbro) [inline crdt/sequence.py:251]: Not sure I understand when self.target would be `None` (vs not) as it is defined in the `__init__` discussion above ("the operation that this operation is performed on") -- could you add an inline comment to clarify?

File: crdt/sequence.py [hunk 12]
```diff
@@ -0,0 +1,247 @@
...
+            # Special case: Append at the end
+            index = len(objects)
+            while index > 0:
+                op = objects[index-1].operation
+                if op.target is None and not self.owner.is_earlier(op.owner):
+                    # We found another operation older than us, so stop iteration.
```
Comment (rebeccabilbro) [inline crdt/sequence.py]: Not sure this is clicking for me -- I might need you add a bit more inline commentary to explain the logic and the need for the `break`

File: tests/test_sequence.py [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+
+        a.remove(3)
+        assert a.get() == ["a", "b", "z"]
+        a.remove(2)
+        assert a.get() == ["a", "b"]
+        a.remove(1)
+        assert a.get() == ["a"]
+        a.remove(0)
+        assert a.get() == []
+
```
Comment (rebeccabilbro) [inline tests/test_sequence.py:72]: add test case for removing from an empty Sequence?

File: tests/test_sequence.py [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+        d.remove(1)
+        assert d.get() == ["x", "y", "z"]
+
+        assert c.merge(d).get() == ["a", "b", "c", "x", "y", "z"]
+        assert d.merge(c).get() == ["a", "b", "c", "x", "y", "z"]
+
+        c.insert(2, "1")
+        assert c.get() == ["a", "b", "1", "c", "x", "y", "z"]
+        assert d.merge(c).get() == ["a", "b", "1", "c", "x", "y", "z"]
+
```
Comment (rebeccabilbro) [inline tests/test_sequence.py:111]: wondering if we want a test for what happens if there are two bobs with conflicting inserts -- so far as I can tell, this is not something we currently support, but it is possible if the names are not required to be unique. maybe we should return an error on `merge` that says something like "sorry, can't reconcile conflicting operations from identically named participants"?

**Review by rebeccabilbro — APPROVED** [review]
This is so cool @pdeziel! I'm excited to see the progress you've made, and it's neat that there is so much of the crdt logic that works at multiple levels, from simple sequences up to the notebook abstraction! 

I've left a handful of comments for your consideration -- one theme is that adding input and return types is probably a good idea if we're expecting to demo this and have external folks exploring the codebase. 

I'm getting excited for our presentation!

File: crdt/sequence.py (hunk 9 above)
Comment (pdeziel) [reply crdt/sequence.py]: Thanks for pointing that out, good to know that exists!

File: crdt/sequence.py (hunk 11 above)
Comment (pdeziel) [reply crdt/sequence.py:251]: When inserting to a sequence with no prior operations applied or at the end there is no target object, because there is nothing to reference against. The current way of handling this is to make `target=None` a switch which means "insert at the end". I've added that info to the docstring.

File: crdt/sequence.py (hunk 12 above)
Comment (pdeziel) [reply crdt/sequence.py]: We need to order all the "insert at the end" operations, which all have `target=None`, so this inserts the operation at the correct position taking into account the normal operation ordering logic. Basically the `target=None` is the equivalent of the nil character in C strings.

File: crdt/sequence.py (hunk 10 above)
Comment (pdeziel) [reply crdt/sequence.py:18]: I do like that idea, I'll create a separate story for it.

File: tests/test_sequence.py (hunk 14 above)
Comment (pdeziel) [reply tests/test_sequence.py:111]: Good call, we currently do not handle this consistently so I added some code which detects this condition and raises an exception, and added a test case for it.
